### PR TITLE
WIP: implemented frame reader/writer

### PR DESCRIPTION
**Pull request by acgreek**
Bug is fixed but want to re-run autobahn to see if there is any improvements in tests

### CODE REVIEW

**Comment by acgreek** [comment]
more tests fail with this than without so we can't merge till we figure those out

File: nbhttp/websocket/conn.go [hunk 1]
```diff
@@ -142,6 +142,17 @@ func (c *Conn) OnMessage(h func(*Conn, MessageType, []byte)) {
 	}
 }
 
+func (c *Conn) OnDataFrame(h func(*Conn, MessageType, bool, []byte)) {
```
Comment (lesismal) [inline nbhttp/websocket/conn.go:145]: Named OnFrameData seems a little better.

File: nbhttp/websocket/conn.go [hunk 2]
```diff
@@ -142,6 +142,17 @@ func (c *Conn) OnMessage(h func(*Conn, MessageType, []byte)) {
...
+	if h != nil {
+		c.dataFrameHandler = func(c *Conn, messageType MessageType, fin bool, data []byte) {
+			c.Server.MessageHandlerExecutor(c.index, func() {
```
Comment (lesismal) [inline nbhttp/websocket/conn.go:148]: the MessageHandlerExecutor should promise the order of the frames, so we need to change it from taskpool.MixedPool to taskpool.FixedPool here:
https://github.com/lesismal/nbio/blob/master/nbhttp/server.go#L242
and
https://github.com/lesismal/nbio/blob/master/nbhttp/server.go#L394

then, we do not need use extra Mutex in OnDataFrame.

File: examples/websocket_tls/sticky/server/server.go [hunk 3]
```diff
@@ -27,9 +28,25 @@ func onWebsocket(w http.ResponseWriter, r *http.Request) {
 		panic(err)
 	}
 	wsConn := conn.(*websocket.Conn)
-	wsConn.OnMessage(func(c *websocket.Conn, messageType websocket.MessageType, data []byte) {
-		// echo
-		c.WriteMessage(messageType, data)
+	mtx := sync.Mutex{}
```
Comment (lesismal) [inline examples/websocket_tls/sticky/server/server.go:31]: no need this Mutex, see
https://github.com/lesismal/nbio/pull/43/files#r661936248

File: examples/websocket_tls/sticky/server/server.go [hunk 4]
```diff
@@ -27,9 +28,25 @@ func onWebsocket(w http.ResponseWriter, r *http.Request) {
...
+	var writer *websocket.LargeMessageWriter
+	wsConn.OnDataFrame(func(c *websocket.Conn, messageType websocket.MessageType, fin bool, data []byte) {
```
Comment (lesismal) [inline examples/websocket_tls/sticky/server/server.go:33]: This example is just effective for an echo service, not useful enough for general functional services.
An exported websocket.Conn.WriteFrame func seems better than LargeMessageWriter:
```golang
func (c *Conn) WriteFrame(messageType MessageType, firstFrame bool, fin bool, data []byte) error {
        return c.writeMessage(messageType, firstFrame, fin, data)
}
```
with WriteFrame, the application layer can easily handle echo message like this:
```golang
var isFirst bool = true
var frames [][]byte
wsConn.OnDataFrame(func(c *websocket.Conn, messageType websocket.MessageType, fin bool, data []byte) {
	err := c.WriteFrame(messageType, isFirst, fin, data)
	if err != nil {
		c.Close()
		return
	}
	isFirst = false
	if fin {
		isFirst = true
	}
})
```
or

```golang
var frames [][]byte
wsConn.OnDataFrame(func(c *websocket.Conn, messageType websocket.MessageType, fin bool, data []byte) {
	frames = append(frames, data)
	if fin {
		var err error
		for i, v := range frames {
			err = c.WriteFrame(messageType, i == 0, i == len(frames)-1, data)
			if err != nil {
				c.Close()
				return
			}
		}
		frames = nil
	}
})
```

File: conn_unix.go [hunk 5]
```diff
@@ -340,10 +337,33 @@ func (c *Conn) write(b []byte) (int, error) {
 		return len(b), nil
 	}
 	c.leftSize += len(b)
-	c.writeBuffers = append(c.writeBuffers, b)
+	c.writeBuffers = append(c.writeBuffers, netmempool.WrapBuffer(b))
```
Comment (lesismal) [inline conn_unix.go:340]: b may from pool.

File: nbhttp/websocket/conn.go (hunk 1 above)
Comment (acgreek) [reply nbhttp/websocket/conn.go:145]: My reason for using OnDataFrame is that the only frames that will be sent to this function are Text and Binary. Ping and Close messages can also have Frame Data, but they will will not passed to this function.

Comment (lesismal) [reply nbhttp/websocket/conn.go:145]: oh yes, I got it.

**Comment by lesismal** [comment]
found the tls pool bug and fixed:
lesismal/llib@8e8f7da#r52993273

please close this and re pr with only the code for the implementation of frame reader/writer

**Pull request closed without merging**